src/Features/VisualBasic/Portable/Structure/InternalStructureBlockStructureProvider.vb

```diff
@@ -0,0 +1,290 @@
+﻿' Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+Imports System.Collections.Immutable
+Imports System.Threading
+Imports Microsoft.CodeAnalysis.Options
+Imports Microsoft.CodeAnalysis.PooledObjects
+Imports Microsoft.CodeAnalysis.Structure
+Imports Microsoft.CodeAnalysis.VisualBasic.Syntax
+
+Namespace Microsoft.CodeAnalysis.VisualBasic.Structure
+    Friend MustInherit Class InternalStructureBlockStructureProvider(Of TBlock As SyntaxNode, THeaderStatement As SyntaxNode, TInnerBlock As SyntaxNode, TPreEndBlock As SyntaxNode, TEndOfBlockStatement As SyntaxNode)
+        Inherits AbstractSyntaxNodeStructureProvider(Of TBlock)
+
+        Friend Sub New()
```

Comment:
this seems much more complex than the C# version.  can you explain why that's necessary?

Reply:
@CyrusNajmabadi
The vbnet implementation is more complex because it is more complex, as the syntax nodes for describing a block structure don't inherit or implement a common interface pattern.

So I've described a potential abstraction (for the outlining providers), it is generic as to reduce the number of type casts required in the provider implementations. I've created a document attempting to describe the abstraction.  [Read Here](https://github.com/AdamSpeight2008/Ideas/blob/master/AbstractInternalStructureProvider.md)